src/signature/signature.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2015-2016 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! Public key signatures: signing and verification.
+//!
+//! Use the `verify` function to verify signatures, passing a reference to the
+//! `_VERIFY` algorithm that identifies the algorithm. See the documentation
+//! for `verify` for examples.
+//!
+//! The design of this module is unusual compared to other public key signature
+//! APIs. Algorithms like split into "signing" and "verification" algorithms.
+//! Also, this API treats each combination of parameters as a separate
+//! algorithm. For example, instead of having a single "RSA" algorithm with a
+//! verification function that takes a bunch of parameters, there are
+//! `RSA_PKCS1_2048_8192_SHA256`, `RSA_PKCS1_2048_8192_SHA512`,
+//! etc. which encode sets of parameter choices into objects. This design is
+//! designed to reduce the risks of algorithm agility. It is also designed to
+//! be optimized for Ed25519, which has a fixed signature format, a fixed curve,
+//! a fixed key size, and a fixed digest algorithm.
+//!
+//! Currently this module does not support digesting the message to be signed
+//! separately from the public key operation, as it is currently being
+//! optimized for Ed25519 and for the implementation of protocols that do not
+//! requiring signing large messages. An interface for efficiently supporting
+//! larger messages will be added later. Similarly, the signing interface is
+//! not available yet.
+
+mod ecdsa;
+mod ed_dsa;
+mod rsa_pkcs1;
```

Comment:
One option here is to add `#[cfg(not(feature = "no_heap"))]` to `rsa_pkcs1` and `ecdsa` instead of having a bunch of these `cfg` feature conditionals inside their respective modules.